tests/test_fetcher.py

```diff
@@ -0,0 +1,89 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of REANA.
+# Copyright (C) 2022 CERN.
+#
+# REANA is free software; you can redistribute it and/or modify it
+# under the terms of the MIT License; see LICENSE file for more details.
+"""REANA-Server workflow fetcher tests."""
+
+import os
+import pytest
+
+from reana_server.fetcher import (
+    _get_github_fetcher,
+    get_fetcher,
+    WorkflowFetcherGit,
+    WorkflowFetcherYaml,
+    WorkflowFetcherZip,
+)
+
+GIT_URL = "https://github.com/reanahub/reana-demo-root6-roofit.git"
+GITHUB_REPO_URL = "https://github.com/reanahub/reana-demo-root6-roofit"
+ZENODO_URL = "https://zenodo.org/record/5752285/files/circular-health-data-processing-master.zip?download=1"
+YAML_URL = "https://raw.githubusercontent.com/reanahub/reana-demo-root6-roofit/master/reana.yaml"
+
+
+@pytest.mark.parametrize(
+    "url, expected_fetcher_class",
+    [
+        (GIT_URL, WorkflowFetcherGit),
+        (GITHUB_REPO_URL, WorkflowFetcherGit),
+        (ZENODO_URL, WorkflowFetcherZip),
+        (YAML_URL, WorkflowFetcherYaml),
+    ],
+)
+def test_fetcher_selection(url, expected_fetcher_class, tmp_path):
+    """Test selection of the fetcher based on the provided URL."""
+    assert isinstance(get_fetcher(url, tmp_path), expected_fetcher_class)
+
+
+@pytest.mark.parametrize(
+    "url, expected_rel_path",
+    [
+        (GIT_URL, "reana.yaml"),
+        (GITHUB_REPO_URL, "reana.yaml"),
+        (GITHUB_REPO_URL + "/blob/master/reana-snakemake.yaml", "reana-snakemake.yaml"),
+        (
+            GITHUB_REPO_URL + "/tree/1fbe52138b98a85a15fde58d385b21b637c18be8",
+            "reana.yaml",
+        ),
+        pytest.param(
+            ZENODO_URL,
+            "reana.yaml",
+            marks=pytest.mark.skip(reason="big file to download"),
```

Comment:
This test is being skipped because it downloads a quite big zip file (~50MB), which might take a while if the connection is not very fast. Do you think we should run this test each time anyway?

Reply:
Mmmh I'm not even sure if we should rely on downloading stuff from those URLs. Even if they should be rather stable, they might change at some point, which would break all the tests. In other words, I'm not sure that unit tests should depend on requesting network resources ie. having an Internet connection at all. Perhaps @tiborsimko has a better-formed opinion about this.

My question was more about, if you skip the test (big zip file is a valid reason, of course), then what's the relevance of keeping this test? Is there any way to force its execution? In other words, under which circumstances that particular test will be executed?

Reply:
I decided to keep this test instead of just omitting it because I think it makes it clear that we are skipping the test only because it is time/bandwidth-consuming and not because of some other reason. In any case, it shouldn't be difficult to rewrite these tests to avoid network calls.

Reply:
What about putting those test zip files etc into `pytest-reana` package and use them here as fixtures instead of downloading them from external URLs?  We have obviously local access to `pytest-reana` when we run tests, so we wouldn't be dependent on any external service, so avoid problems you mentioned above such as (i) network resource troubles but also (ii) what-if-author-changes-zenodo-record troubles.

If our test zip file is small, i.e. we delete most of the regular source file content and keep only `reana.yaml` there, then it will nicely fit into PyPI package as `include_package_data`, so it could be practical solution. And we could mock URL loading from Zenodo but fetch the file from there.

TL;DR WDYT about treating examples as other pytest fixtures?

Reply:
I think it's not only zip files, but all that's hosted  behind these links:

```
GIT_URL = "https://github.com/reanahub/reana-demo-root6-roofit.git"
GITHUB_REPO_URL = "https://github.com/reanahub/reana-demo-root6-roofit"
ZENODO_URL = "https://zenodo.org/record/5752285/files/circular-health-data-processing-master.zip?download=1"
YAML_URL = "https://raw.githubusercontent.com/reanahub/reana-demo-root6-roofit/master/reana.yaml"
```

If I'm not mistaken, all this content is downloaded to perform the tests.